libraries/core_libs/consensus/include/transaction/transaction_manager.hpp

```diff
@@ -246,6 +247,7 @@ class TransactionManager : public std::enable_shared_from_this<TransactionManage
   std::unordered_map<trx_hash_t, std::shared_ptr<Transaction>> nonfinalized_transactions_in_dag_;
   std::unordered_map<trx_hash_t, std::shared_ptr<Transaction>> recently_finalized_transactions_;
   std::unordered_map<PbftPeriod, std::vector<trx_hash_t>> recently_finalized_transactions_per_period_;
+  ExpirationCacheMap<trx_hash_t, uint64_t> estimations_cache_;
```

Comment:
I think you need to have cache tx_gas per period. Contract calls might have very different gas estimations in different periods ?